source/core/shape.cc

```diff
@@ -0,0 +1,428 @@
+/* ---------------------------------------------------------------------
+ *
+ * Copyright (C) 2019 - by the Lethe authors
+ *
+ * This file is part of the Lethe library
+ *
+ * The Lethe library is free software; you can use it, redistribute
+ * it, and/or modify it under the terms of the GNU Lesser General
+ * Public License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ * The full text of the license can be found in the file LICENSE at
+ * the top level of the Lethe distribution.
+ *
+ * ---------------------------------------------------------------------
+ */
+#include <core/shape.h>
+
+template <int dim>
+double
+Shape<dim>::displaced_volume(const double fluid_density)
+{
+  StandardExceptions::ExcNotImplemented();
+  return 1.;
+}
+
+template <int dim>
+Point<dim>
+Shape<dim>::align_and_center(const Point<dim> &evaluation_point) const
+{
+  // Translation and rotation to standard position and orientation for
+  // distance calculations
+  Point<dim> center_of_rotation = position;
+
+  Point<dim> rotated_point;
+  Point<dim> translated_point;
+
+  // Rotation from the solid orientation
+  // Angular position around x, y and z axis
+  Tensor<1, 3> theta = orientation;
+
+  // The centralized point is the one to be rotated, and it is updated after
+  // each rotation around one axis The centralized rotated point is the result
+  // of each rotation, and it is initialized in case no rotation is performed
+  Point<dim> centralized_point;
+  centralized_point              = evaluation_point - center_of_rotation;
+  Point<dim> centralized_rotated = centralized_point;
+
+  // Selection of the first axis around which to rotate:
+  // x -> 0, y -> 1, z -> 2
+  // In 2D, only rotation around the z axis is possible
+  if (dim == 2)
```

Comment:
can't this be a const expr? because dim is known at compile time you don't need a regular if